### PR TITLE
Added High-level Architecture of Control Diagram to /Docs

### PR DESCRIPTION
**Pull request by EvanMorcom**
### Description
Added High-level Architecture of Control Diagram to /Docs.

This diagram shows the main components of the control stack and how they interact with each other. This diagram was created to help explain the stack and gather feedback.

### CODE REVIEW

**Comment by EvanMorcom** [comment]
@garethellis0 @MathewMacDougall @akhilveeraghanta @jonathanlew @ZhikaiZhang1 @jennaslo Please review the diagram and let me know if there are changed I should make to clarify 👍

**Comment by EvanMorcom** [comment]
Screen Shot 2020-05-27 at 12 44 08 PM

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Nice diagram! 
Nit: rurebook->rulebook
Can you add documentation for how the Python integration will work? Is that just used to generate C code?

**Review by ZhikaiZhang1 — COMMENTED** [review]
"Thanks for the diagram! Will there also be diagram for the wheel controller or is that already up in the repo?"

Yeah I'll make more diagrams for the parts of the stack that need more explanation 👍

**Comment by MathewMacDougall** [comment]
Only think I don't fully understand is the "Polynomial Position Error compensation". Specifically, what's the polynomial part?

**Comment by EvanMorcom** [comment]
> Only think I don't fully understand is the "Polynomial Position Error compensation". Specifically, what's the polynomial part?

That's the part where the robot will re-plan a polynomial to get it back on the main trajectory if it leaves the path or accumulates too much error before the software navigator replans. I've updated the block to explain this better.

**Comment by EvanMorcom** [comment]
> Nice diagram!
> Nit: rurebook->rulebook
> Can you add documentation for how the Python integration will work? Is that just used to generate C code?

Thanks, fixed the typo 👍 

The python code is just for design essentially. The controller will be generated to a header at compile time, so I don't think it fits in this diagram as it's focused more along the flow of control.

**Comment by EvanMorcom** [comment]
Screen Shot 2020-05-27 at 7 37 06 PM

**Comment by EvanMorcom** [comment]
Screen Shot 2020-05-27 at 7 39 32 PM

**Review by akhilveeraghanta — COMMENTED** [review]
image lgtm, should we put it in an `md` file titled `trajectory_planner_and_controller.md` or something along those lines, with a description?

**Comment by EvanMorcom** [comment]
> image lgtm, should we put it in an `md` file titled `trajectory_planner_and_controller.md` or something along those lines, with a description?

Yep 👍 

I'm going to include it in a markdown doc with more explanation, just thought I would gather feedback about the diagram first

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Some suggestions:
- "Controller generated compile time" -> "generated from python model at compile-time"
- Not sure if it makes sense for the controller to have an arrow back to `AI Layer`, really it's causing changes in robot state that are picked up by the camera and fed back to the AI
- Is there some way we can indicate that the firmware and software are both using the same trajectory planner?

**Comment by garethellis0** [comment]
Also as a note, you can view the image under "Files Changed" if you change the view from raw text to the formatted version, no need to post it as a comment.

**Review by jonathanlew — COMMENTED** [review]
>     * Not sure if it makes sense for the controller to have an arrow back to `AI Layer`, really it's causing changes in robot state that are picked up by the camera and fed back to the AI

I agree, I think that robot state should be its own box

**Comment by EvanMorcom** [comment]
> > ```
> > * Not sure if it makes sense for the controller to have an arrow back to `AI Layer`, really it's causing changes in robot state that are picked up by the camera and fed back to the AI
> > ```
> 
> 
> I agree, I think that robot state should be its own box

Added a vision box to clarify where the state updates are coming from

**Review by MathewMacDougall — DISMISSED** [review]
LGTM

**Review by jonathanlew — COMMENTED** [review]
> The python code is just for design essentially. The controller will be generated to a header at compile time, so I don't think it fits in this diagram as it's focused more along the flow of control.

Is the Python only to generate the wheel controllers?

> Added a vision box to clarify where the state updates are coming from

Can we complete the loop i.e., `wheel controller->robot physics->vision`? Also, wouldn't the onboard controller also use the encoders since that will be faster than vision?

**Comment by EvanMorcom** [comment]
> > The python code is just for design essentially. The controller will be generated to a header at compile time, so I don't think it fits in this diagram as it's focused more along the flow of control.
> 
> Is the Python only to generate the wheel controllers?
> 
> > Added a vision box to clarify where the state updates are coming from
> 
> Can we complete the loop i.e., `wheel controller->robot physics->vision`? Also, wouldn't the onboard controller also use the encoders since that will be faster than vision?

Yep I can add the robot physics :+1:. 

Yeah the wheel controls have their loop closed with the encoder data to correct their velocity. The vision feedback is used by the algorithm for re-planning the XY path

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm :tada:

**Comment by MathewMacDougall** [comment]
`PythonControl.svg`
* The "simulated trajectory" line between the `Robot Simulator` and `Performance Metrics` isn't straight :grimacing: :sweat_smile: 

`ControlStock.svg`
* I think you mean `ControlStack.svg`? (stock -> stack)

`TrajectoryPlanner.svg`
* In order to calculate forwards and backwards continuity, you're calculating velocity and acceleration since that's the "limiting factor" of what the robot can do physically. Because of that it makes sense to calculate the velocity-based trajectory first and then operate on that. However the decision to do most manipulation on the position-based trajectory is to minimize error, correct? (Due to operating at lower derivative, so less error when integrating etc.)

**Comment by EvanMorcom** [comment]
> `PythonControl.svg`
> 
> * The "simulated trajectory" line between the `Robot Simulator` and `Performance Metrics` isn't straight 😬 😅
> 
> `ControlStock.svg`
> 
> * I think you mean `ControlStack.svg`? (stock -> stack)
> 
> `TrajectoryPlanner.svg`
> 
> * In order to calculate forwards and backwards continuity, you're calculating velocity and acceleration since that's the "limiting factor" of what the robot can do physically. Because of that it makes sense to calculate the velocity-based trajectory first and then operate on that. However the decision to do most manipulation on the position-based trajectory is to minimize error, correct? (Due to operating at lower derivative, so less error when integrating etc.)

Straightened out the line 👍 

I'll figure out how to rename the file through draw.io.

We could technically do the entire trajectory without calculating the position trajectory and go straight to x,y,w velocities. But, the position trajectory is used at a high level as it is what we are interested in. The velocity trajectory is used at the wheel level and the position at the robot/ai level.

File: PythonControl.svg [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
```
Comment (garethellis0) [inline PythonControl.svg]: Are we passing `Simulated Trajectory` back to the `Wheel Controller` from `Robot Simulator`? Wouldn't we just be passing the current state?

File: docs/images/ControlStock.svg [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
..."translate(-0.5 -0.5)"><switch><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe flex-start; width: 148px; height: 1px; padding-top: 650px; margin-left: 612px;"><div style="box-sizing: border-box; font-size: 0; text-align: left; "><div style="display: inline-block; font-size: 12px; font-family: Helvetica; color: #000000; line-height: 1.2; pointer-events: all; white-space: normal; word-wrap: normal; ">Note:<br />Trajectory Planner is identical in software and firmware but is run on both</div></div></div></foreignObject><text x="612" y="654" fill="#000000" font-family="Helvetica" font-size="12px">Note:...</text></switch></g></g><switch><g requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"/><a transform="translate(0,-5)" xlink:href="https://desk.draw.io/support/solutions/articles/16000042487" target="_blank"><text text-anchor="middle" font-size="10px" x="50%" y="100%">Viewer does not support full SVG 1.1</text></a></switch></svg>
```
Comment (garethellis0) [inline docs/images/ControlStock.svg]: For both the `Navigator` stack and `Trajectory Planner stack, might it make more sense to represent them as boxes around their constituent parts? Right now it seems like a cycle that can repeat (which it is not, unless I'm misunderstanding something).

Comment (garethellis0) [inline docs/images/ControlStock.svg]: I like how you denoted what `Desired Robot State` actually is right below. Could we do something similar for the other things we're passing around? At least to indicate what space each path/trajectory is in.

File: docs/images/TrajectoryPlanner.svg [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
...orm="translate(-0.5 -0.5)"><switch><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 118px; height: 1px; padding-top: 110px; margin-left: 1391px;"><div style="box-sizing: border-box; font-size: 0; text-align: center; "><div style="display: inline-block; font-size: 12px; font-family: Helvetica; color: #000000; line-height: 1.2; pointer-events: all; white-space: normal; word-wrap: normal; ">Linear Interpolation of Constant Time segments</div></div></div></foreignObject><text x="1450" y="114" fill="#000000" font-family="Helvetica" font-size="12px" text-anchor="middle">Linear Interpolation...</text></switch></g></g><switch><g requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"/><a transform="translate(0,-5)" xlink:href="https://desk.draw.io/support/solutions/articles/16000042487" target="_blank"><text text-anchor="middle" font-size="10px" x="50%" y="100%">Viewer does not support full SVG 1.1</text></a></switch></svg>
```
Comment (garethellis0) [inline docs/images/TrajectoryPlanner.svg]: Could we add a side-note about what `Balance Segment Duration` means? That really confused me until you explained it.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good! Just a few comments for added detail / readability.

File: PythonControl.svg (hunk 1 above)
Comment (EvanMorcom) [reply PythonControl.svg]: What was intended here is to show that the performance metric are used to change the wheel controller. I've updated the diagram to reflect this. The diagram is more to show how the controller is designed and updated the C firmware than the operation of the controller.

File: docs/images/ControlStock.svg (hunk 2 above)
Comment (EvanMorcom) [reply docs/images/ControlStock.svg]: I've updated the trajectory planner stack to be a single box in the firmware side. I think it's important to show the iterative nature of the navigator stack optimizing a path.

File: docs/images/TrajectoryPlanner.svg (hunk 3 above)
Comment (EvanMorcom) [reply docs/images/TrajectoryPlanner.svg]: Added 👍

File: docs/images/TrajectoryPlanner.svg [hunk 4]
```diff
@@ -0,0 +1,3 @@
...
...41" width="40" height="20" fill="none" stroke="none" pointer-events="all"/><g transform="translate(-0.5 -0.5)"><switch><foreignObject style="overflow: visible; text-align: left;" pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 451px; margin-left: 940px;"><div style="box-sizing: border-box; font-size: 0; text-align: center; "><div style="display: inline-block; font-size: 12px; font-family: Helvetica; color: #000000; line-height: 1.2; pointer-events: all; white-space: nowrap; ">Text</div></div></div></foreignObject><text x="940" y="455" fill="#000000" font-family="Helvetica" font-size="12px" text-anchor="middle">Text</text></switch></g></g><switch><g requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"/><a transform="translate(0,-5)" xlink:href="https://desk.draw.io/support/solutions/articles/16000042487" target="_blank"><text text-anchor="middle" font-size="10px" x="50%" y="100%">Viewer does not support full SVG 1.1</text></a></switch></svg>
```
Comment (jonathanlew) [inline docs/images/TrajectoryPlanner.svg]: random text box below the diagram

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: docs/images/TrajectoryPlanner.svg (hunk 4 above)
Comment (EvanMorcom) [reply docs/images/TrajectoryPlanner.svg]: Whoops, fixed

**Review by jonathanlew — APPROVED** [review]
🖼️ 🔢

**Review by MathewMacDougall — APPROVED** [review]
LGTM pending Gareth's remaining comments

**Review by garethellis0 — APPROVED** [review]

**Comment by EvanMorcom** [comment]
Failed to DL again, I'm going to restart again and see what happens

**Pull request merged**